### ISSUE TITLE
Track when in_app_include is used

### ISSUE THREAD

**Issue by armenzg** [getsentry/sentry-python#3671]
If the customer sets an `in_app_include`, we may destroy it with one of our enhancement rules.

Recently we noticed that the customer installed their code under their virtual env which ends up under `site-packages` which we then mark as **not an in-app frame** ([path:\*\*/site-packages/\*\* -app](https://github.com/getsentry/sentry/blob/e56c063b7c55de625354f07114896a3fdb308d65/src/sentry/grouping/enhancer/enhancement-configs/newstyle%402023-01-11.txt#L377)).

You can see the effects of it here (a few of the 12 more frames should have been in-app):
![Image](https://github.com/user-attachments/assets/1a04ef2f-ace4-437d-bb8c-708a99fce73d)

The replacements can be in the associated JSON:
```json
"abs_path": "/venv-path/site-packages/foo/code.py",
"in_app": false,
"data": { "orig_in_app": 1 }
```

From the Sentry POV, we need to know which frames were modified by the `in_app_include` rules to prevent the values from being overwritten.

If we added a key saying `in_app_include_used` and the value of the rule we would then prevent the overwriting and perhaps even show it in the UI what `in_app_include` sets the value.

What do you think?

CC @antonpirker

**Comment by sentrivana** [comment]
If someone sets `in_app_include` explicitly this should definitely be respected in Sentry. `in_app_include_used` with the pattern that matched sounds like a good idea to me. 👍🏻

**Comment by sentrivana** [comment]
Looked into this some more, trying to figure out why the decision from the SDK is sometimes overridden (i.e., cannot be relied on) in the first place and if we can improve that instead.

Sentry overrides the decision for frames that come from `site-packages` and `dist-packages`, i.e., if the SDK marks them as `in_app`, they are then set to `in_app = False` by Sentry. I believe we can remove this rule since the SDK already checks this (see https://github.com/getsentry/sentry-python/blob/336b17714c8101c5f3896915b37acbb8bca5f3fa/sentry_sdk/utils.py#L1059).

So for now we won't be adding a new field and will instead look into aligning the decision between the SDK and Sentry. This includes (1) removing the duplicate `site/dist-packages` check on the Sentry side as well as (2) looking into examples of questionable `in_app` decisions from the SDK.

**Comment by lobsterkatie** [comment]
> Looked into this some more, trying to figure out why the decision from the SDK is sometimes overridden (i.e., cannot be relied on) in the first place and if we can improve that instead.
> 
> Sentry overrides the decision for frames that come from `site-packages` and `dist-packages`, i.e., if the SDK marks them as `in_app`, they are then set to `in_app = False` by Sentry. I believe we can remove this rule since the SDK already checks this

I agree, having the same logic in both the SDK and the server seems a bit redundant, and we can probably only doing the checks in one place. I'd actually argue that place should be the server, though, rather than the SDK, since that's where it happens for any language which needs sourcemapping/symbolication/etc, and it'd be nice to have the server just be where it happens for everyone.

That said, regardless of where we do the auto in-app/out-of-app checks, there's always going to be the possibility of in-app value changes on the server, because users can mark things in or out of app with stacktrace enhancement rules. 

To keep things aligned with custom fingerprinting (which can also happen both in the client and on the server, with the server taking precedence), I think we should preserve the server-beats-client-ness when comparing two custom in-app decisions. So really the only case where things currently aren't right is when a server auto-decision beats a client custom decision. So here's my new proposal:

- In the client, mark `in-app` as `None`, unless overridden by user config. Don't do any auto-marking.
- On the server, we'll have the following logic:
  - Incoming value is `None` -> can be overridden by both an auto decision and a custom stacktrace rule
  - Incoming value is set -> can only be overridden by a stacktrace rule
  - Server stacktrace rules beat server auto-decisions.

What do you think?

**Comment by sentrivana** [comment]
@lobsterkatie Thanks for taking the time to write this down!

The proposal makes sense to me. I agree the logic should live on the server and the SDK should just mark frames if they're in the user's `in_app_include`/`in_app_exclude` set on SDK init.

My only concern with this is: Can we break anything by removing this from the SDK -- i.e., does anything else rely on `in_app` being set by the SDK? As far as I can tell from a quick grep through Sentry that isn't the case so hopefully we're good there.

So just to make sure we're on the same page, this would be the plan from the SDK point of view:
- Mark a frame as `in_app = True` iff it comes from a module from the user's `in_app_include` (takes precedence over `exclude`)
- Mark a frame as `in_app = False` iff it comes from a module from the user's `in_app_exclude`
- If neither of the above applies, set `in_app = None`

If this sounds good I'll whip up an SDK PR for this tomorrow. (And clean up the issues since https://github.com/getsentry/sentry-python/issues/3682 is then stale.)

**Comment by lobsterkatie** [comment]
Sorry for totally losing track of this - I've now picked it back up as my current main task. Yes, if you could make that PR for the SDK, that would be great. On my end, I'm going to:

- Write a new issue to track this, to supersede all of the other issues/PRs where conversation has happened.
- Make sure our server-side logic for auto-marking a frame as in or out of app matches that currently in the SDK, so we don't lose any of that logic when you remove it. At that point the SDK change can be merged.
- Fix the server-side logic to match what I wrote above.

Sound good?

**Comment by sentrivana** [comment]
Yup! Followed up in https://github.com/getsentry/sentry/issues/83603
